### PR TITLE
Oauth2 support

### PR DESCRIPTION
**Pull request by willyborankin**
Add support for OAuth2 auth

### CODE REVIEW

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java [hunk 1]
```diff
@@ -156,6 +146,106 @@ public boolean visible(final String name, final Map<String, Object> parsedConfig
...
+                CONNECTION_GROUP,
+                groupCounter++,
+                ConfigDef.Width.LONG,
+                OAUTH2_CLIENT_SCOPE_CONFIG
+        );
+        configDef.define(
+                OAUTH2_RESPONSE_TOKEN_PROPERTY_CONFIG,
+                ConfigDef.Type.STRING,
+                "access_token",
+                new ConfigDef.Validator() {
```
Comment (HelenMel) [inline src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:233]: Why don't some default validators like 'nonEmptyString' doesn't work?

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java [hunk 2]
```diff
@@ -156,6 +146,106 @@ public boolean visible(final String name, final Map<String, Object> parsedConfig
...
+                ConfigDef.Width.LONG,
+                OAUTH2_CLIENT_AUTHORIZATION_MODE_CONFIG
+        );
+        configDef.define(
+                OAUTH2_CLIENT_SCOPE_CONFIG,
+                ConfigDef.Type.STRING,
+                null,
+                new ConfigDef.NonEmptyStringWithoutControlChars(),
+                ConfigDef.Importance.LOW,
+                "The scope used for fetching access token.",
```
[the remark this answers is not in the document]
Comment (ivanyu) [reply src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:223]: Yep, Kafka itself uses short sentences with simplified grammar (no dots, no articles) for the human-readable title and normal punctuation & grammar (with dots, complete sentences, articles) for the doc strings ([example](https://github.com/apache/kafka/blob/954c090ffc378a63ce3c3c9a72b87724fcd2cd6c/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java)). Let's stick to this

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java [hunk 3]
```diff
@@ -287,11 +394,76 @@ public final String connectorName() {
...
+
+    public final String oauth2ClientScope() {
+        return getString(OAUTH2_CLIENT_SCOPE_CONFIG);
+    }
+
+    public final String oauth2ResponseTokenProperty() {
+        return getString(OAUTH2_RESPONSE_TOKEN_PROPERTY_CONFIG);
+    }
+
+    private static class UrlValidator implements ConfigDef.Validator {
```
Comment (HelenMel) [inline src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:429]: Lets move it to  separate file

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java [hunk 4]
```diff
@@ -287,11 +394,76 @@ public final String connectorName() {
...
     }
 
+    public final URI oauth2AccessTokenUri() {
+        return getString(OAUTH2_ACCESS_TOKEN_URL_CONFIG) != null ? toURI(OAUTH2_ACCESS_TOKEN_URL_CONFIG) : null;
+    }
+
+    private URI toURI(final String propertyName) {
+        try {
+            return new URL(getString(propertyName)).toURI();
+        } catch (final MalformedURLException | URISyntaxException e) {
```
Comment (HelenMel) [inline src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:404]: Could it actually happend? I think we checked this value with URL Validator?

File: src/main/java/io/aiven/kafka/connect/http/sender/AccessTokenHttpRequestBuilder.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                                      final StringJoiner requestBodyBuilder,
+                                      final HttpSinkConfig config) {
+        if (config.oauth2AuthorizationMode() == OAuth2AuthorizationMode.HEADER) {
+            final var clientAndSecretBytes =
+                    (config.oauth2ClientId() + ":" + config.oauth2ClientSecret().value())
+                            .getBytes(StandardCharsets.UTF_8);
+            final var clientAndSecretAuthHeader =
+                    "Basic " + Base64.getEncoder().encodeToString(clientAndSecretBytes);
+            httpRequestBuilder.header(HEADER_AUTHORIZATION, clientAndSecretAuthHeader);
+        } else {
```
Comment (HelenMel) [inline src/main/java/io/aiven/kafka/connect/http/sender/AccessTokenHttpRequestBuilder.java:65]: Lets explicitly specify that config.oauth2AuthorizationMode() == OAuth2AuthorizationMode.URL. It make intention a bit more clear

File: src/main/java/io/aiven/kafka/connect/http/sender/OAuth2HttpSender.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    //for testing
+    OAuth2HttpSender(final HttpSinkConfig config, final HttpClient httpClient) {
+        super(config, httpClient);
+    }
+
+    @Override
+    protected void sendWithRetries(final HttpRequest.Builder requestBuilder,
+                                   final HttpResponseHandler originHttpResponseHandler) {
+        final HttpResponseHandler handler = response -> {
+            if (response.statusCode() == 401) { // access denied or refresh of a token is needed
```
Comment (HelenMel) [inline src/main/java/io/aiven/kafka/connect/http/sender/OAuth2HttpSender.java:56]: Implementation is absolutely fine, but I would rather make some scheduled task to refresh tokens and do not interrupt a flow.

Comment (ivanyu) [reply src/main/java/io/aiven/kafka/connect/http/sender/OAuth2HttpSender.java:56]: I wonder if it's really worth to increase the complexity...

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java (hunk 4 above)
Comment (willyborankin) [reply src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:404]: well it cannot, but `MalformedURLException` and `URISyntaxException` are checked exceptions. `URL` and `URI` constructors throw theme explicitly.

**Comment by willyborankin** [comment]
@HelenMel and @tvainika all fixed comments in a separate commit

**Review by HelenMel — APPROVED** [review]

**Review by HelenMel — APPROVED** [review]

**Pull request merged**